include/aos/common/tools/list.hpp

```diff
@@ -0,0 +1,616 @@
+/*
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef AOS_LIST_HPP_
+#define AOS_LIST_HPP_
+
+#include <assert.h>
+
+#include "aos/common/tools/buffer.hpp"
+#include "aos/common/tools/error.hpp"
+#include "aos/common/tools/new.hpp"
+
+namespace aos {
+
+template <typename T>
+class List;
+
+/**
+ * List implementation.
+ * @tparam T list type.
+ */
+template <typename T>
+class ListImpl {
+protected:
+    struct Node {
+        bool  mAllocated = false;
+        T     mValue;
+        Node* mNext = nullptr;
+        Node* mPrev = nullptr;
+    };
+
+    /**
+     * List iterator.
+     */
+    class Iterator {
```

Comment:
It seems like the difference between ConstIterator & Iterator is in dereference operator.
A single template should be enough to cover both cases

Reply:
Redesigned.